### PR TITLE
Implement sort order

### PR DESCRIPTION
**Pull request by ericzolf**
I'm almost done but I have one small issue: I don't manage to have the logview redraw when the tasks list changes (especially when the order changes).

i.e. I'd expect something like Alt-2 (grouped) -> Alt-0 (sort by task order) -> Ctrl-T (change the task order) to automatically change the task order in the LogView pane, but I need to press again e.g. Alt-0 to make it happen

I also still need to update the shortcuts pane, and probably documentation and test cases, but wanted a first opinion.

### CODE REVIEW

**Comment by ericzolf** [comment]
OK, documentation and shortcuts were quickly done.

File: src/gtimelog/data/org.gtimelog.gschema.xml
Comment (mgedmin) [inline src/gtimelog/data/org.gtimelog.gschema.xml:12]: I'd be slightly happier with `as_is` or `as-is`.

Or we could be explicit and go with `chronological`?

Or I could stop nitpicking and learn to live with `asis`, which I help but pronounce as a single word.

File: src/gtimelog/main.py
Comment (mgedmin) [inline src/gtimelog/main.py:1349]:
```suggestion
        blurb='Log order of tasks/groups (asis/alphanum/duration/tasks)')
```

File: src/gtimelog/main.py
Comment (mgedmin) [inline src/gtimelog/main.py:1469]: I'm somewhat surprised that this setting takes effect only in grouped view, and not also in summary view.

File: src/gtimelog/main.py
Comment (mgedmin) [inline src/gtimelog/main.py:1524]: list.index() can raise an IndexError if there's no match.  Also, it does an O(N) linear search over the entire list.

I think it'd be better to keep a dictionary mapping tasks to ordering, and then do

    return lambda x: self.task_order.get(x[1], sys.maxsize)

to put all unknown tasks at the end.

Another thing, how does this handle ties?  E.g. what about two tasks that took exactly the same time, when sorting by duration?  How should they be sorted?

File: src/gtimelog/main.py
Comment (mgedmin) [inline src/gtimelog/main.py:1529]: This is probably the place to build that auxiliary sort structure.

Note to self: review how task list updates are being handled, i.e. will this be called every time the content of the task list change, or do we need to register an observer somewhere?

File: src/gtimelog/timelog.py
Comment (mgedmin) [inline src/gtimelog/timelog.py:1084]: There should be a blank line between paragraphs of text inside docstrings.

Also, I think instead of "unsorted" it should say "in source order".

(Any reason it couldn't be named `tasks`?  or `all_tasks`?)

File: src/gtimelog/timelog.py
Comment (mgedmin) [inline src/gtimelog/timelog.py:1146]: Okay, this is a bit of a trap for code reviewers then.  It looks like the list.index() or str.index(), but behaves differently.

I'd like it to be be renamed.  Maybe `order()`?

File: src/gtimelog/main.py
Comment (mgedmin) [inline src/gtimelog/main.py:1390]: For this to work you have to define a property `tasks` in this class (which is LogView, right?)  See how class TaskListView does it:

```python
    tasks = GObject.Property(
        type=object, nick='Tasks',
        blurb='The task list (an instance of TaskList)')
```

And then instead of a `self.log_view.set_tasks(tasks)`, bind the property in `Window.__init__`, where it currently binds Window.tasks to TaskListView.tasks:

```python
        self.bind_property('tasks', self.task_list, 'tasks', GObject.BindingFlags.DEFAULT)    # <-- existing code
        self.bind_property('tasks', self.log_view, 'tasks', GObject.BindingFlags.DEFAULT)    # <-- new code
```

Then the assignment to `self.tasks` in Window.load_tasks will automatically invoke `LogView.queue_update`.

File: src/gtimelog/main.py
Comment (mgedmin) [reply src/gtimelog/main.py:1524]: Okay, so self.tasks is a TaskList and not a list, and `index()` is a custom helper.  Disregard most of the above.

The part about handling ties still stands.

File: src/gtimelog/main.py
Comment (mgedmin) [reply src/gtimelog/main.py:1529]: Actually no, this method should just go away.

File: src/gtimelog/timelog.py
Comment (mgedmin) [reply src/gtimelog/timelog.py:1084]: New, better suggestion: instead of keeping a list, build a dictionary from task name to its order in the original file.  Then reimplement `def index()` (renamed to `def order()`) to do a `return self.task_order.get(task, sys.maxsize)`.

File: src/gtimelog/timelog.py
Comment (mgedmin) [inline src/gtimelog/timelog.py:1127]: This `for line in f:` would become `for index, line in enumerate(f)` and you'd do `task_order[line] = index` here.

**Review by mgedmin — COMMENTED** [review]
I like most of what I see, but I also have a bunch of suggestions.

The mystery of the missing update remains.  I didn't notice anything wrong in the code.  Have you played with the GTK Inspector and checked whether the various components get their properties updated in time?  Added debug prints to populate_log()?

**Comment by mgedmin** [comment]
It's fine to ignore coverage for now, while the structure of the code is still up in the air.

File: src/gtimelog/main.py
Comment (mgedmin) [reply src/gtimelog/main.py:1524]: I think I want these sort helpers to be global functions in the timelog.py module, so they can have unit tests.  Then this function can become

```python
        return {
            'alphanum': sort_by_entry_title,
            'duration': sort_by_duration,
            'tasks': sort_by_task_order(self.tasks),
            'asis': sort_by_start_time,
        }[self.log_order]
```

There's a slight asymmetry here where sort_by_task_order() is a higher order function, but I don't think it can be helped:

```python
def sort_by_entry_title(item):
    start, entry, duration = item
    return (entry, start, duration)

def sort_by_duration(item):
    start, entry, duration = item
    return (duration, start, entry)

def sort_by_start_time(item):
    return item

def sort_by_task_order(tasks):

    def sort_key(item):
        start, entry, duration = item
        return (tasks.order(entry), item)

    return sort_key
```

I've also made it clearer what these x[1], x[2] mean and added tie breaking (by falling back to the previous sort rules, which are start time, then entry title, then duration).

I should probably remember this is the 21st century and refactor TimeCollection.grouped_entries() to return named tuples.

File: src/gtimelog/main.py
Comment (ericzolf) [reply src/gtimelog/main.py:1349]: Taken.

File: src/gtimelog/main.py
Comment (ericzolf) [reply src/gtimelog/main.py:1469]: To be honest, I got lazy, and it would have been another completely different implementation, because the structures are so different, but I can probably do it.

File: src/gtimelog/timelog.py
Comment (ericzolf) [reply src/gtimelog/timelog.py:1146]: Renamed and description adapted.

File: src/gtimelog/timelog.py
Comment (ericzolf) [reply src/gtimelog/timelog.py:1127]: Yep. done.

File: src/gtimelog/data/org.gtimelog.gschema.xml
Comment (mgedmin) [reply src/gtimelog/data/org.gtimelog.gschema.xml:12]: I like it.

File: src/gtimelog/main.py
Comment (mgedmin) [reply src/gtimelog/main.py:1469]: It's not a merge blocker.  If you don't do it, I think I will.

File: src/gtimelog/main.py
Comment (ericzolf) [reply src/gtimelog/main.py:1529]: Removed.

File: src/gtimelog/main.py
Comment (ericzolf) [reply src/gtimelog/main.py:1469]: OK, thanks.

File: src/gtimelog/main.py
Comment (ericzolf) [reply src/gtimelog/main.py:1390]: Aha, I couldn't find any conclusive documentation. Thanks for the explanations. Implemented and works.

File: src/gtimelog/main.py
Comment (ericzolf) [reply src/gtimelog/main.py]: I would rename your functions in accordance with the new options name start-time/name/duration/task-list (or the other way around if you prefer).

Also, it's still a bit fuzzy, but I'm asking myself if `_get_log_order_key` shouldn't become a method of "window/TimeCollection", so that we simply call `work, slack = window.grouped_entries(sorted_by=self.log_order, sorted_tasks=self.tasks)` and let `grouped_entries` completely decide what it does with the information. This is slightly more complex, _but_ more encapsulated also, and it would make it easier to keep the same interface later on for `categorized_work_entries`.

Comment (mgedmin) [reply src/gtimelog/main.py]: Sure, that works for me.

Comment (ericzolf) [reply src/gtimelog/main.py]: Result is slightly different, but I think the spirit is the same. I've tried to capture my thoughts in the comments of the code in timelog. Next step: add tests...

**Comment by ericzolf** [comment]
I'm starting to wonder where are the tests giving a 100% coverage of TimeCollection, where I only find few tests in `src/gtimelog/tests/test_timelog.py`. Looking at testing `_get_grouped_order_key`.

**Comment by ericzolf** [comment]
All checks have passed, PR ready for final review. Do you want me now to squash my commits, or do you do it yourself?

File: src/gtimelog/tests/test_timelog.py
Comment (mgedmin) [inline src/gtimelog/tests/test_timelog.py:1046]: Using tuples here feels a bit weird.  Why not lists?

File: src/gtimelog/tests/test_timelog.py
Comment (mgedmin) [inline src/gtimelog/tests/test_timelog.py:1051]: For readability purposes I'd prefer `[entry for start_time, entry, duration in sorted(...)]`

File: src/gtimelog/timelog.py
Comment (mgedmin) [inline src/gtimelog/timelog.py:370]: I'm not sure what "for the ordered function" refers to.

```suggestion
        """Return a callable usable as the `key` argument of sorted().
```

File: src/gtimelog/timelog.py
Comment (mgedmin) [inline src/gtimelog/timelog.py:376]: Oh, good catch.

File: src/gtimelog/timelog.py
Comment (mgedmin) [inline src/gtimelog/timelog.py:384]: Hm.  My gut feeling says that start-time is better here (followed by duration for tie breaking), but I haven't thought about it very long.

**Review by mgedmin — APPROVED** [review]
Mostly LGTM.  There are possible improvements, but I won't feel too bad merging it as is if you'd rather not spend any more time on it.

**Comment by mgedmin** [comment]
I don't think squashing is necessary.  The actual history may be interesting to look at.

**Comment by mgedmin** [comment]
Oh, it would also be helpful if you could help me come up with a short description of this feature for the changelog (in CHANGES.rst).

File: src/gtimelog/main.py
Comment (ericzolf) [reply src/gtimelog/main.py:965]: Then, removed, and indeed, no difference noticed.

File: src/gtimelog/menus.ui
Comment (ericzolf) [reply src/gtimelog/menus.ui:114]: None other than inattention, aligned labels between menu and dialog, _and_ aligned with the target (e.g. "By name" for "name").

File: src/gtimelog/tests/test_timelog.py
Comment (ericzolf) [reply src/gtimelog/tests/test_timelog.py:1046]: Because I always use tuples where the "list" doesn't change after. I'm assuming that it has some (minimal) performance advantage.

File: src/gtimelog/timelog.py
Comment (ericzolf) [reply src/gtimelog/timelog.py:370]: Done.

File: src/gtimelog/timelog.py
Comment (ericzolf) [reply src/gtimelog/timelog.py:384]: As my main ask for this whole thing was to have a sorting stable across weeks/days, the name, after the task-list itself, was for me the better choice (as commented).

**Comment by ericzolf** [comment]
> I don't think squashing is necessary. The actual history may be interesting to look at.

Matter of taste, your Git repo, no issue! I'm a "squasher" :nerd_face: , one topic per PR, one commit per PR, smaller/cleaner `git log` (but less history of course, you can't have everything :shrug:).

> Oh, it would also be helpful if you could help me come up with a short description of this feature for the changelog (in CHANGES.rst).

I added an entry to the changelog.

**Comment by mgedmin** [comment]
> Matter of taste, your Git repo, no issue! I'm a "squasher" nerd_face , one topic per PR, one commit per PR.

My personal preference used to be fast-forward merges after heavy rebasing into small self-contained commits that each change one thing only.  And then GitHub happened, with the big shiny green Merge button, and I gave up on fast-forward merges.

**Review by mgedmin — APPROVED** [review]

**Comment by mgedmin** [comment]
Thank you!

**Comment by mgedmin** [comment]
I almost forgot the most important bit: how would you like to be credited in CONTRIBUTORS.rst?

**Comment by ericzolf** [comment]
Eric Lavarde is fine.

**Pull request merged**
